banyand/property/repair.go

```diff
@@ -0,0 +1,467 @@
+		if next == nil {
+			return nil
+		}
+		var shaValue string
+		var source []byte
+		var deleteTime int64
+		for err == nil && next != nil {
+			hitNumber = next.HitNumber
+			var errTime error
+			err = next.VisitStoredFields(func(field string, value []byte) bool {
+				switch field {
+				case shaValueField:
+					shaValue = convert.BytesToString(value)
+				case sourceField:
+					source = value
+				case deleteField:
+					deleteTime = convert.BytesToInt64(value)
+				}
+				return true
+			})
+			if err = multierr.Combine(err, errTime); err != nil {
+				return errors.WithMessagef(err, "visit stored fields, hit: %d", hitNumber)
+			}
+			err = each(source, shaValue, deleteTime)
+			if err != nil {
+				return errors.WithMessagef(err, "processing source failure, hit: %d", hitNumber)
+			}
+			latestDocValues = next.SortValue
+			next, err = result.Next()
+		}
+	}
+}
+
+func (r *repair) buildShaValue(source []byte, deleteTime int64) (string, error) {
+	var err error
+	hash := sha512.New()
+	_, err = hash.Write(source)
+	if err != nil {
+		return "", fmt.Errorf("hashing source failure: %w", err)
+	}
+	_, err = hash.Write([]byte(fmt.Sprintf("%d", deleteTime)))
+	if err != nil {
+		return "", fmt.Errorf("hashing delete time failure: %w", err)
+	}
+	return fmt.Sprintf("%x", hash.Sum(nil)), nil
+}
+
+func (r *repair) readState() (*repairStatus, error) {
+	stateFile, err := os.ReadFile(r.statePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	var status repairStatus
+	err = json.Unmarshal(stateFile, &status)
+	if err != nil {
+		return nil, err
+	}
+	return &status, nil
+}
+
+type repairStatus struct {
+	Trees          map[string]*repairTree `json:"trees"`
+	LastSnapshotID uint64                 `json:"last_snapshot_id"`
+}
+
+type repairTree struct {
+	Root *repairTreeNode `json:"root"`
+}
+
+type repairTreeNode struct {
+	ShaValue string            `json:"sha_value"`
+	ID       string            `json:"id"`
+	Children []*repairTreeNode `json:"children"`
+}
+
+type snapshotIDList []uint64
+
+func (s snapshotIDList) Len() int           { return len(s) }
+func (s snapshotIDList) Less(i, j int) bool { return s[i] < s[j] }
+func (s snapshotIDList) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+type repairTreeComposer struct {
+	file      *os.File
+	path      string
+	slotCount int
+}
+
+func newRepairTreeComposer(path string, slotCount int) (r *repairTreeComposer, err error) {
+	if path == "" {
+		return nil, fmt.Errorf("repair tree file path cannot be empty")
+	}
+	r = &repairTreeComposer{
+		path:      path,
+		slotCount: slotCount,
+	}
+	r.file, err = os.OpenFile(r.path, os.O_CREATE|os.O_RDWR, storage.FilePerm)
+	if err != nil {
+		return nil, fmt.Errorf("opening repair tree file %s failure: %w", r.path, err)
+	}
+	return
+}
+
+func (r *repairTreeComposer) append(group, id, shaVal string) (err error) {
+	if err = r.writeString(group); err != nil {
```

Comment:
You can use `encoding.EncodeBytes` to encode string.

Reply:
If I use `encoding.EncodeBytes`, It needs one more `int32` space to save the length of the byte array.

Reply:
`encoding.EncodeBytes` encodes the length in bytes automatically. You don't need to encode the length separately, which simplifies your code.

Reply:
But I need do decode the specific length when reading?

Reply:
You can use the DecodeBytes from the same pkg